### PR TITLE
Feature/more providers

### PR DESCRIPTION
**Pull request by Lansoweb**
Provide a narrative description of what you are trying to accomplish:

- [ ] Are you fixing a bug?
  - [ ] Detail how the bug is invoked currently.
  - [ ] Detail the original, incorrect behavior.
  - [ ] Detail the new, expected behavior.
  - [ ] Base your feature on the `master` branch, and submit against that branch.
  - [ ] Add a regression test that demonstrates the bug, and proves the fix.
  - [ ] Add a `CHANGELOG.md` entry for the fix.

- [X] Are you creating a new feature?
  - [X] Why is the new feature needed? What purpose does it serve?
    This allows the usage of other providers, like GitLab and BitBucket (in the future)
  - [X] How will users use the new feature?
    One can use the --provider option to commands.
  - [X] Base your feature on the `develop` branch, and submit against that branch.
  - [X] Add only one feature per pull request; split multiple features over multiple pull requests
  - [ ] Add tests for the new feature.
  - [X] Add documentation for the new feature.
  - [X] Add a `CHANGELOG.md` entry for the new feature.

- [ ] Is this related to quality assurance?
  

- [ ] Is this related to documentation?

### CODE REVIEW

File: src/AddEntry.php [hunk 1]
```diff
@@ -72,6 +72,10 @@ public function __invoke(string $type, string $changelogFile, string $entry) : v
     /**
      * Locates the location within the changelog where the injection should occur.
      * Also determines if the injection is a replacement or an addition.
+     *
+     * @param array $contents
+     * @param string $type
+     * @return stdClass
```
Comment (weierophinney) [inline src/AddEntry.php]: param and return annotations are not necessary when typehints exist, unless you add descriptions.

File: src/EntryCommand.php [hunk 2]
```diff
@@ -17,7 +19,7 @@
 
 class EntryCommand extends Command
 {
-    use GetChangelogFileTrait;
+    use GetChangelogFileTrait, GetProviderTrait;
```
Comment (weierophinney) [inline src/EntryCommand.php]: One `use` statement per line, please.

File: src/AddEntry.php (hunk 1 above)
Comment (Lansoweb) [reply src/AddEntry.php]: True. Updating ...

Comment (weierophinney) [reply src/AddEntry.php]: (I'll take care of these during merge.)

Comment (Lansoweb) [reply src/AddEntry.php]: Actually, this happened in other files too. My IDE automatically fill these ... Update them all?

File: src/Provider/GitLab.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+declare(strict_types=1);
+
+namespace Phly\KeepAChangelog\Provider;
+
+use Gitlab\Client as GitLabClient;
+
+class GitLab implements ProviderInterface
+{
+    const PACKAGE_NAME = 'm4tthumphrey/php-gitlab-api';
```
Comment (weierophinney) [inline src/Provider/GitLab.php]: I don't see this constant referenced anywhere; what is it for?

File: src/Provider/GitLab.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * @inheritDoc
+     */
```
Comment (weierophinney) [inline src/Provider/GitLab.php]: Considering that the docblocks for the interface only duplicate the information in the typehints, we can remove those, and then remove these `@inheritDoc` entries. Alternately, give _descriptions_ to the docblocks in the interface.

File: src/TaggerCommand.php [hunk 5]
```diff
@@ -97,17 +104,4 @@ protected function execute(InputInterface $input, OutputInterface $output) : int
 
         return 0;
     }
-
-    private function tagWithChangelog(string $tagName, string $package, string $version, string $changelog) : bool
```
Comment (weierophinney) [inline src/TaggerCommand.php]: I think the tagging should likely be the same between different providers; what will differ is _releasing_. We may also want to vary what message is generated for the tag, as I noted earlier.

As such, I'd argue that the interface should remove the `createLocalTag()` method, and that should move back to this class. We can then vary the message created, using either `--no-tag-summary` or `--short` or similar.

File: src/Provider/GitLab.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    public function createLocalTag(string $tagName, string $package, string $version, string $changelog) : bool
+    {
+        $command = sprintf('git tag -s -m "%s %s" %s', $package, $version, $tagName);
```
[the remark this answers is not in the document]
Comment (Lansoweb) [reply src/Provider/GitLab.php]: Yes. The release procedure is different between providers. In GitLab, the tag message is plain text and showed in a single line (as in [here](https://gitlab.com/LeandroSilva/changelog-test/tags)), only `release notes` use markdown.

**Comment by Lansoweb** [comment]
PR updated reverting tag behaviour and removing unnecessary methods comments. Sorry for the delay.

**Pull request merged**